### PR TITLE
ref(hc): Only set the organizationmembermapping.member_id for monolith

### PR DESCRIPTION
**Pull request by corps**
Support lookup in invite with orgmember_id only for regions that were historically monolith.

Simplify all existing organizationmembermapping logic.

### CODE REVIEW

File: src/sentry/receivers/outbox/region.py [hunk 1]
```diff
@@ -94,15 +96,22 @@ def process_organization_member_updates(
 
     rpc_org_member_update = RpcOrganizationMemberMappingUpdate.from_orm(org_member)
 
-    organizationmember_mapping_service.update_with_organization_member(
-        organizationmember_id=org_member.id,
-        organization_id=shard_identifier,
-        rpc_update_org_member=rpc_org_member_update,
-    )
+    if _was_monolith():
```
Comment (markstory) [inline src/sentry/receivers/outbox/region.py:99]: How do non-US regions update membership mappings?

Comment (corps) [reply src/sentry/receivers/outbox/region.py:99]: Oh, right, what I need to do is simply remove the organizationmember_id=org_member.id, parameter, not the entire update, and write a test for that. Good catch.

Comment (corps) [reply src/sentry/receivers/outbox/region.py:99]: Oh, right, what I need to do is simply remove the organizationmember_id=org_member.id, parameter, not the entire update, and write a test for that. Good catch.

Comment (corps) [reply src/sentry/receivers/outbox/region.py]: Took a different approach -- take another pass?

File: src/sentry/api/endpoints/accept_organization_invite.py [hunk 2]
```diff
@@ -41,9 +44,25 @@ def get_invite_state(
         user_id: int,
     ) -> Optional[RpcUserInviteContext]:
         if organization_slug is None:
-            member_mapping = OrganizationMemberMapping.objects.filter(
-                organizationmember_id=member_id
-            ).first()
+            # Lookup by the organization member id, but only for historical monolith regions.
+            member_mapping: OrganizationMemberMapping | None = None
```
Comment (corps) [inline src/sentry/api/endpoints/accept_organization_invite.py:48]: My original thinking was to only set `organizationmember_id` for regions that are monolith.  I realized this made updates and deletes not work.

New approach -- invites work for monolith regions by looking up the organizations associated with all the member mappings associated with an id, and finding the one that is associated with the monolith region.

File: src/sentry/api/endpoints/organization_member/index.py [hunk 3]
```diff
@@ -286,9 +285,7 @@ def post(self, request: Request, organization) -> Response:
             if settings.SENTRY_ENABLE_INVITES:
                 om.token = om.generate_token()
             om.save()
-            region_outbox = om.save_outbox_for_create()
```
Comment (corps) [inline src/sentry/api/endpoints/organization_member/index.py:289]: Outboxes are upserts by default now.

File: src/sentry/services/hybrid_cloud/organizationmember_mapping/impl.py [hunk 4]
```diff
@@ -20,84 +19,53 @@
 
 
 class DatabaseBackedOrganizationMemberMappingService(OrganizationMemberMappingService):
-    def create_mapping(
+    def upsert_mapping(
```
Comment (corps) [inline src/sentry/services/hybrid_cloud/organizationmember_mapping/impl.py:22]: Greatly simplify all this logic with better upsert behavior.

File: src/sentry/services/hybrid_cloud/organizationmember_mapping/model.py [hunk 5]
```diff
@@ -25,14 +25,6 @@ class RpcOrganizationMemberMapping(RpcModel):
 
 
 class RpcOrganizationMemberMappingUpdate(RpcModel):
-    """
```
Comment (corps) [inline src/sentry/services/hybrid_cloud/organizationmember_mapping/model.py:28]: This comment is longer true -- it's a true replacement call.

File: src/sentry/api/endpoints/organization_member/details.py [hunk 6]
```diff
@@ -181,13 +181,10 @@ def put(
...
                 if result.get("regenerate"):
                     if request.access.has_scope("member:admin"):
-                        region_outbox = None
                         with transaction.atomic():
                             member.regenerate_token()
                             member.save()
-                            region_outbox = member.save_outbox_for_update()
-                        if region_outbox:
-                            region_outbox.drain_shard(max_updates_to_drain=10)
+                        member.outbox_for_update().drain_shard(max_updates_to_drain=10)
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_member/details.py:187]:
```suggestion 
                        # Use outbox_for_update() to build an unsaved outbox message that we can use to drain with
                        member.outbox_for_update().drain_shard(max_updates_to_drain=10)
```

This was confusing on first read, as it looks like we're missing a save() call on the outbox message. I realized afterwards that you're using `outbox_for_update()` to rebuild the attributes necessary to drain the messages, perhaps we could include a comment for future us?

**Comment by corps** [comment]
@markstory 
> Invites are only valid for 30 days. We could start minting new invitation emails/links using organization slugs now, and by the time we're ready to onboard customers the US and EU regions can behave the same as all the valid invites will already have organization slugs in them.

Awesome, let's consider this for a future change that will simplify many things.

File: src/sentry/services/hybrid_cloud/organizationmember_mapping/model.py (hunk 5 above)
Comment (RyanSkonnord) [reply src/sentry/services/hybrid_cloud/organizationmember_mapping/model.py:28]: Looks like it hasn't been true for some time. An "omitted key" is possible with a `TypedDict` but not a Pydantic model. Did this get changed from a `TypedDict` at some point?

**Review by markstory — APPROVED** [review]

**Review by dashed — APPROVED** [review]

**Review by vbro — APPROVED** [review]

**Pull request merged**